crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -5889,11 +5889,19 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.infer_binary_expression_type(binary.into(), false, left_ty, right_ty, *op)
             .unwrap_or_else(|| {
                 if let Some(builder) = self.context.report_lint(&UNSUPPORTED_OPERATOR, binary) {
-                    builder.into_diagnostic(format_args!(
+                    let mut diag = builder.into_diagnostic(format_args!(
                         "Operator `{op}` is unsupported between objects of type `{}` and `{}`",
                         left_ty.display(self.db()),
                         right_ty.display(self.db())
                     ));
+
+                    if op == &ast::Operator::BitOr && left_ty.into_class_literal().is_some()
+                        && right_ty.into_class_literal().is_some() && Program::get(self.db()).python_version(self.db())
```

Comment:
I'm not sure we need to check the RHS type at all here. There could be cases like `int | foo` where `foo` is something other than a class literal type where this runtime error would still occur, and this hint would be just as useful.

Reply:
Yes, if either the l.h.s. or the r.h.s. is a subtype of `type`, it should be fine to display the hint. `TypeVar("T") | int` succeeds on Python 3.10+ but not on earlier versions